csrc/iter_visitor.h

```diff
@@ -372,4 +370,146 @@ class TORCH_CUDA_CU_API InputsOf : public IterVisitor {
       const std::vector<Val*>& outputs_);
 };
 
+//! This is a generic traversal class that is used to modify a Fusion graph by
+//! replacing Vals to simplify computation or remove dead code. This differs
+//! from OptOutMutator, which is built for mutating TensorViews in-place in a
+//! graph by altering the associated IterDomains, and which does not easily
+//! handle modifying TensorView definitions and Fusion outputs during traversal.
+//!
+//! Derived classes should override handle() for relevant Exprs and they should
+//! make use of replaceVal() to change the definitions of Vals in the graph.
+//! Note that if replacements are made using replaceVal(old_val, new_val), then
+//! neither new_val nor any new Statements produced in creating it will be
+//! traversed by this class. Also note that any Vals or Exprs that are
+//! previously marked dead will not be processed by handle().
+class DeadCodeRemover : BackwardVisitor {
+ public:
+  DeadCodeRemover(Fusion* fusion) : BackwardVisitor(false), fusion_(fusion) {}
+
+  DeadCodeRemover(const DeadCodeRemover& other) = default;
+  DeadCodeRemover& operator=(const DeadCodeRemover& other) = default;
+
+  DeadCodeRemover(DeadCodeRemover&& other) = default;
+  DeadCodeRemover& operator=(DeadCodeRemover&& other) = default;
+
+  //! Instead of traverseTo, run() is the entry point for this class, and we
+  //! always traverse from outputs backward to their inputs.
+  void run();
+
+  inline Fusion* fusion() const {
+    return fusion_;
+  }
+
+ protected:
+  using BackwardVisitor::handle;
+
+  void handle(Statement* stmt) override;
+  void handle(Expr* expr) override;
+
+  //! We implement this in order to remove dangling TensorViews whose uses are
+  //! all dead. Note that we do not remove other ValTypes like Scalars since
+  //! they might still be used as attributes or members of other objects, which
+  //! is not reflected by Val::uses().
+  void handle(TensorView* tv) override;
+
+  //! Replaces a Val in outputs, and in all uses.
+  //!
+  //! The argument old_val is always marked Dead by this method. If old_val is a
+  //! Fusion input, we do not replace it. If old_val's definition is non-null
+  //! and has other outputs which are not dead, we do not remove old_val.
+  //!
+  //! Returns whether old_val was registered for removal from the Fusion.
+  bool replaceVal(Val* old_val, Val* new_val);
```

Comment:
Please mention this does not do the replacement right away but registers the replacement.